src/components/PostDetails.tsx

```diff
@@ -4,163 +4,142 @@ import { NewCommentForm } from './NewCommentForm';
   const [visible, setVisible] = useState(false);
 
-  function loadComments() {
-    setLoaded(false);
-    setError(false);
-    setVisible(false);
-
-    commentsApi
-      .getPostComments(post.id)
-      .then(setComments) // save the loaded comments
-      .catch(() => setError(true)) // show an error when something went wrong
-      .finally(() => setLoaded(true)); // hide the spinner
-  }
-
-  useEffect(loadComments, [post.id]);
-
-  // The same useEffect with async/await
-  /*
-  async function loadComments() {
-    setLoaded(false);
-    setVisible(false);
-    setError(false);
-
-    try {
-      const commentsFromServer = await commentsApi.getPostComments(post.id);
-
-      setComments(commentsFromServer);
-    } catch (error) {
-      setError(true);
-    } finally {
-      setLoaded(true);
+  useEffect(() => {
+    if (!selectedPost) {
+      return;
     }
-  };
 
-  useEffect(() => {
-    loadComments();
-  }, []);
+    dispatch(setLoaded(false));
+    dispatch(setError(false));
+    setVisible(false);
 
-  useEffect(loadComments, [post.id]); // Wrong!
-  // effect can return only a function but not a Promise
-  */
+    commentsApi
+      .getPostComments(selectedPost.id)
+      .then(commentsFromServer => {
+        dispatch(setComments(commentsFromServer));
+      })
+      .catch(() => dispatch(setError(true)))
+      .finally(() => dispatch(setLoaded(true)));
+  }, [dispatch, selectedPost]);
 
   const addComment = async ({ name, email, body }: CommentData) => {
     try {
       const newComment = await commentsApi.createComment({
         name,
         email,
         body,
-        postId: post.id,
+        postId: selectedPost?.id || 0,
       });
 
-      setComments(currentComments => [...currentComments, newComment]);
-
-      // setComments([...comments, newComment]);
-      // works wrong if we wrap `addComment` with `useCallback`
-      // because it takes the `comments` cached during the first render
-      // not the actual ones
+      dispatch(setComment(newComment));
     } catch (error) {
-      // we show an error message in case of any error
-      setError(true);
+      dispatch(setError(true));
     }
   };
 
   const deleteComment = async (commentId: number) => {
-    // we delete the comment immediately so as
-    // not to make the user wait long for the actual deletion
-    // eslint-disable-next-line max-len
-    setComments(currentComments =>
-      currentComments.filter(comment => comment.id !== commentId),
-    );
+    dispatch(removeComment(commentId));
 
     await commentsApi.deleteComment(commentId);
   };
 
   return (
     <div className="content" data-cy="PostDetails">
-      <div className="block">
-        <h2 data-cy="PostTitle">{`#${post.id}: ${post.title}`}</h2>
-
-        <p data-cy="PostBody">{post.body}</p>
-      </div>
+      {selectedPost && (
+        <>
+          <div className="block">
+            <h2 data-cy="PostTitle">{`#${selectedPost.id}: ${selectedPost.title}`}</h2>
```

Comment:
Consider using destruction assignment for `selectedPost` object and creating separate component for this